### DISCUSSION TITLE
Is it possible to return SignInActivity properties using PnP.PowerShell?

### DISCUSSION THREAD

**Discussion by cprovpo** [pnp/powershell#3541] [Q&A] [answered]
Hello!  I currently am trying to use Get-PnPAzureADUser to return the Property SignInActivity, which includes items like LastSignInDateTime.  However, it does not appear this is available from the list of available properties
![image](https://github.com/pnp/powershell/assets/139259660/f4977550-e231-4546-8d60-46fe10aa8339)

I am currently getting this information in a script that leverages the Microsoft.Graph module (Get-MgUser).  However, a rising number of conflicts between PnP.PowerShell and Microsoft.Graph has me looking to move as many scripts to PnP-only as possible.

Is it possible to return SignInActivity properties using PnP.PowerShell?

**Comment by AndersRask** [comment]
What version of PnP are you using? The latest builds have fixed a lot of the issues with dll collisions since it uses ALC bridging pattern.

https://github.com/pnp/powershell/issues/3207

You only get the basic properties from Get-PnPAzureADUser unfortunately.

**Reply by gautamdsheth** [reply]
You can specify additional properties like mentioned here:

https://pnp.github.io/powershell/cmdlets/Get-PnPAzureADUser.html#example-5

**Reply by AndersRask** [reply]
Yes but those does not include the properties he is asking for though. I couldn't even find the SignInActivity properties in Az, so I think it is a Mg specific property collection

**Reply by gautamdsheth** [reply]
If you request any additional Graph properties via `Select` parameter, they will be retrieved and should be available in `AdditionalProperties`

**Reply by AndersRask** [reply]
Great! I did this poc and it seems to work:

```
$additionalProps = Get-PnPAzureADUser -Identity "someuser@mytenant.onmicrosoft.com" -select SignInActivity | select -expand AdditionalProperties 
$additionalProps.GetEnumerator() | select -expand value

```
@gautamdsheth do you know if it is planned to support advanced queries in Get-PnPAzureADUser?
It is now supported in Mg and Az using -ConsistencyLevel Eventual -Count
https://learn.microsoft.com/en-us/graph/aad-advanced-queries?tabs=http

Example:
 Get-PnPAzureADUser -Filter "endsWith(mail,'hotmail.com')"
currently fails with unsupported query and I have found no way to set Header

**Reply by cprovpo** [reply]
@AndersRask  - Thanks for your POC code, but it didn't look like it could drill down to actual date/time values for that property.

```
$additionalProps

Key            Value
---            -----
signInActivity {lastSignInDateTime, lastSignInRequestId, lastNonInteractiveSignInDateTime, lastNonInteractiveSignInRequestId}

$values = $additionalProps.getEnumerator() | select -expand Value 

HasValues    : False
Type         : String
Parent       : {}
Root         : {lastSignInDateTime, lastSignInRequestId, lastNonInteractiveSignInDateTime, lastNonInteractiveSignInRequestId}
Next         : 
Previous     : 
Path         : lastSignInDateTime
First        : 
Last         : 
LineNumber   : 0
LinePosition : 0

HasValues    : False
Type         : String
Parent       : {}
Root         : {lastSignInDateTime, lastSignInRequestId, lastNonInteractiveSignInDateTime, lastNonInteractiveSignInRequestId}
Next         : 
Previous     : 
Path         : lastSignInRequestId
First        : 
Last         : 
LineNumber   : 0
LinePosition : 0

HasValues    : False
Type         : String
Parent       : {}
Root         : {lastSignInDateTime, lastSignInRequestId, lastNonInteractiveSignInDateTime, lastNonInteractiveSignInRequestId}
Next         : 
Previous     : 
Path         : lastNonInteractiveSignInDateTime
First        : 
Last         : 
LineNumber   : 0
LinePosition : 0

HasValues    : False
Type         : String
Parent       : {}
Root         : {lastSignInDateTime, lastSignInRequestId, lastNonInteractiveSignInDateTime, lastNonInteractiveSignInRequestId}
Next         : 
Previous     : 
Path         : lastNonInteractiveSignInRequestId
First        : 
Last         : 
LineNumber   : 0
LinePosition : 0

```

According to Get-Module, the object that is returned here is a Newtonsoft.Json.Linq.JProperty.  Even with all properties selected from Value, this is the entirety of the output:

```
Name                       : lastSignInDateTime
Type                       : Property
HasValues                  : True
First                      : {}
Last                       : {}
Count                      : 1
Parent                     : {lastSignInDateTime, lastSignInRequestId, lastNonInteractiveSignInDateTime, lastNonInteractiveSignInRequestId}
Root                       : {lastSignInDateTime, lastSignInRequestId, lastNonInteractiveSignInDateTime, lastNonInteractiveSignInRequestId}
Next                       : {}
Previous                   : 
Path                       : lastSignInDateTime
LineNumber                 : 0
LinePosition               : 0
IsReadOnly                 : False
AllowNew                   : True
AllowEdit                  : True
AllowRemove                : True
SupportsChangeNotification : True
SupportsSearching          : False
SupportsSorting            : False
IsSorted                   : False
SortProperty               : 
SortDirection              : Ascending
IsFixedSize                : False
SyncRoot                   : System.Object
IsSynchronized             : False
```

@gautamdsheth - Am I misunderstanding how to pull the actual property values here?

**Reply by AndersRask** [reply, marked as answer]
Yeah Newtonsoft Json is a bit hard to figure out in PowerShell for sure.
A little "hack" is to convert the Dictionary object you get back from additional properties to json and *back*, that will give you a PSObject that is easier to manage :-)
Also remember that your output may vary depending on if you get 1 or more logins back

```
$properties = $additionalProps | ConvertTo-Json | ConvertFrom-Json
$properties.signInActivity.lastSignInDateTime
```

**Reply by cprovpo** [reply]
That did the trick!  Thank you!

For anyone else that encounters this issue, it's a combination of answers from @AndersRask and @gautamdsheth that looks like this:

```
$additionalProps = Get-PnPAzureADUser -Identity "user@domain.com" -select SignInActivity | select -expand AdditionalProperties 
$additionalProps.GetEnumerator() | select -expand value
$properties = $additionalProps | ConvertTo-Json | ConvertFrom-Json
$properties.signInActivity.lastSignInDateTime
```

**Comment by cprovpo** [comment]
> What version of PnP are you using? The latest builds have fixed a lot of the issues with dll collisions since it uses ALC bridging pattern.

@AndersRask  - I'm actually using the following:

PnP.PowerShell - 2.2.77 Nightly Build
Microsoft.Graph 2.8.0

Even today, after trying Start-ThreadJob, I'm still getting errors like the one below because PnP loads up the older version of Microsoft.Graph.Core before I call the Microsft.Graph cmdlets.

```
Get-MgUser_List: 
Line |
  10 |          $AllUsers = Get-MgUser -All -Property $Properties #| Select-O …
     |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Could not load type 'Microsoft.Graph.Authentication.AzureIdentityAccessTokenProvider' from assembly 'Microsoft.Graph.Core, Version=1.25.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
```

**Reply by AndersRask** [reply]
What version of powershell are you using?

**Reply by cprovpo** [reply]
```
Name                           Value
----                           -----
PSVersion                      7.3.9
```

**Reply by AndersRask** [reply]
Ok that is strange then! It should work without the thread jobs even, as the assemblies should be loaded "on demand".
Maybe file this as a seperate issue in issue list with code to reproduce. Looks like a bug
